File: workspace/solutions/kotlin/lib/src/main/kotlin/s0/s9/s0/s7/s_0907_00/s_0907_00.kt
/*
 * @lc app=leetcode.cn id=907 lang=kotlin
 *
 * [907] 子数组的最小值之和
 */

package s0.s9.s0.s7.s_0907_00

// @lc code=start
class Solution {
    companion object {
        val MOD = 1_000_000_007L
    }

    fun sumSubarrayMins(arr: IntArray): Int {
        val n = arr.size
        var stack = mutableListOf<Int>()
        var left = IntArray(n)
        var right = IntArray(n)

        arr.forEachIndexed { i, item ->
            while (stack.isNotEmpty() && arr[stack.last()] >= item) {
                stack.removeLast()
            }
            left[i] = stack.lastOrNull() ?: -1
            stack.add(i)
        }

        stack.clear()
        arr.reversed().forEachIndexed { i, item ->
            val j = n - i - 1
            while (stack.isNotEmpty() && arr[stack.last()] > item) {
                stack.removeLast()
            }
            right[j] = stack.lastOrNull() ?: n
            stack.add(j)
        }

        var result = 0L
        for (i in arr.indices) {
            result = (result + (i - left[i]).toLong() * (right[i] - i) * arr[i]) % MOD
        }
        return result.toInt()
    }
}
// @lc code=end

File: workspace/solutions/ecmascript/src/s0/s9/s0/s7/s_0907_00.js
/*
 * @lc app=leetcode.cn id=907 lang=javascript
 *
 * [907] 子数组的最小值之和
 */

// @lc code=start
const MOD = 1e9 + 7;
/**
 * @param {number[]} arr
 * @return {number}
 */
const sumSubarrayMins = function (arr) {
    const n = arr.length;
    const stack = [];
    const left = Array.from({ length: n });
    const right = Array.from({ length: n });

    for (let i = 0; i < n; i++) {
        // @ts-ignore
        while (stack.length > 0 && arr[stack.at(-1)] >= arr[i]) {
            stack.pop();
        }
        left[i] = stack.at(-1) ?? -1;
        stack.push(i);
    }

    stack.length = 0;
    for (let i = n - 1; i >= 0; i--) {
        // @ts-ignore
        while (stack.length > 0 && arr[stack.at(-1)] > arr[i]) {
            stack.pop();
        }
        right[i] = stack.at(-1) ?? n;
        stack.push(i);
    }

    let result = 0;
    for (let i = 0; i < n; i++) {
        // @ts-ignore
        result += arr[i] * (i - left[i]) * (right[i] - i);
        result %= MOD;
    }
    return result;
};
// @lc code=end

export default sumSubarrayMins;
